temporalio/workflow.py

```diff
@@ -0,0 +1,2026 @@
+import temporalio.exceptions
+
+WorkflowClass = TypeVar("WorkflowClass", bound=Type)
+ExternalWorkflowClass = TypeVar("ExternalWorkflowClass")
+LocalParamType = TypeVar("LocalParamType")
+ActivityReturnType = TypeVar("ActivityReturnType")
+WorkflowReturnType = TypeVar("WorkflowReturnType")
+T = TypeVar("T")
+
+
+@overload
+def defn(cls: WorkflowClass) -> WorkflowClass:
+    ...
+
+
+@overload
+def defn(*, name: str) -> Callable[[WorkflowClass], WorkflowClass]:
+    ...
+
+
+def defn(cls: Optional[WorkflowClass] = None, *, name: Optional[str] = None):
+    """Decorator for workflow classes.
+
+    This must be set on any registered workflow class (it is ignored if on a
+    base class).
+
+    Args:
+        cls: The class to decorate.
+        name: Name to use for the workflow. Defaults to class ``__name__``.
+    """
+
+    def with_name(name: str, cls: WorkflowClass) -> WorkflowClass:
+        # This performs validation
+        _Definition._apply_to_class(cls, name)
+        return cls
+
+    # If name option is available, return decorator function
+    if name is not None:
+        return partial(with_name, name)
+    if cls is None:
+        raise RuntimeError("Cannot create defn without class or name")
+    # Otherwise just run decorator function
+    return with_name(cls.__name__, cls)
+
+
+WorkflowRunFunc = TypeVar("WorkflowRunFunc", bound=Callable[..., Awaitable[Any]])
+
+
+def run(fn: WorkflowRunFunc) -> WorkflowRunFunc:
+    """Decorator for the workflow run method.
+
+    This must be set on one and only one async method defined on the same class
+    as ``@workflow.defn``. This can be defined on a base class method but must
+    then be explicitly overridden and defined on the workflow class.
+
+    Run methods can only have positional parameters. Best practice is to only
+    take a single object/dataclass argument that can accept more fields later if
+    needed.
+
+    Args:
+        fn: The function to decorate.
+    """
+    if not inspect.iscoroutinefunction(fn):
+        raise ValueError("Workflow run method must be an async function")
+    # Disallow local classes because we need to have the class globally
+    # referenceable by name
+    if "<locals>" in fn.__qualname__:
+        raise ValueError(
+            "Local classes unsupported, @workflow.run cannot be on a local class"
+        )
+    setattr(fn, "__temporal_workflow_run", True)
+    return fn
+
+
+WorkflowSignalFunc = TypeVar(
+    "WorkflowSignalFunc", bound=Callable[..., Union[None, Awaitable[None]]]
+)
+
+
+@overload
+def signal(fn: WorkflowSignalFunc) -> WorkflowSignalFunc:
+    ...
+
+
+@overload
+def signal(*, name: str) -> Callable[[WorkflowSignalFunc], WorkflowSignalFunc]:
+    ...
+
+
+@overload
+def signal(
+    *, dynamic: Literal[True]
+) -> Callable[[WorkflowSignalFunc], WorkflowSignalFunc]:
+    ...
+
+
+def signal(
+    fn: Optional[WorkflowSignalFunc] = None,
+    *,
+    name: Optional[str] = None,
+    dynamic: Optional[bool] = False,
```

Comment:
Could simplify with special value: `name: Union[Literal[CatchAll, str]]` and get rid of `dynamic`..

Reply:
Not really, because this is the final destination of the overloads. And one overload is `name=` and the other is `dynamic=`. So I _have_ to accept both parameters in the final target method because it has to be a superset of overload parameters. Unless you are saying `@workflow.signal(name=CatchAll)` is more user-friendly `@workflow.signal(dynamic=True)`.

Note, non-`@overload` functions are not documentation in API doc. This is basically the best way to do mutually exclusive parameters.

Reply:
I was thinking `@workflow.signal(name=CatchAll)` is more friendly, it also implies that you can combine named and catchall signal handlers.

Reply:
I disagree it's more friendly (and it'd be `@workflow.signal(name=workflow.CatchAll)`). They are different mutually exclusive parameters. Also, you can't combine them because the dynamic catch all one is given a name and then the args (we actually enforce at runtime that they accept self, a name, then positional varargs).